src/instructions.rs

```diff
@@ -0,0 +1,814 @@
+    } else {
+        // Since the immediate flag was off, we only need the SR2 section (first 3 bits).
+        // This section contains the register containing the value to add.
+        let sr2 = Register::from_u16(instr & THREE_BIT_MASK)?;
+        regs[dr] = regs[sr1].wrapping_add(regs[sr2]);
+    }
+
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Does the bitwise 'NOT' for a value stored in a register.
+///
+/// ### Arguments
+///
+/// - `instr`: An u16 that has the encoding of the whole instruction to execute.
+/// - `regs`: A Registers struct that handles each register.
+pub fn not(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    let dr = Register::from_u16((instr >> 9) & THREE_BIT_MASK)?;
+    let sr = Register::from_u16((instr >> 6) & THREE_BIT_MASK)?;
+
+    regs[dr] = !regs[sr];
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Does the bitwise 'AND' between two values and stores the result
+/// in a register.
+///
+/// This instruction can happen in two different ways. One is
+/// by doing it with two elements that are in a register each,
+/// this is called register mode. The other mode is to use the value
+/// of a register with the one that is embedded in the instruction itself.
+///
+/// ### Arguments
+///
+/// - `instr`: An u16 that has the encoding of the whole instruction to execute.
+/// - `regs`: A Registers struct that handles each register.
+pub fn and(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    // Destination register
+    let dr = Register::from_u16((instr >> 9) & THREE_BIT_MASK)?;
+    // SR1 section
+    let sr1 = Register::from_u16((instr >> 6) & THREE_BIT_MASK)?;
+    // Imm flag
+    let imm_flag = (instr >> 5) & ONE_BIT_MASK;
+
+    if imm_flag == 1 {
+        // Get the imm5 section, then do the bitwise and with the content on R1.
+        let mut imm5 = instr & FIVE_BIT_MASK;
+        imm5 = sign_extend(imm5, 5)?;
+        regs[dr] = regs[sr1] & imm5;
+    } else {
+        // Get the SR2 section, then do the bitwise and with the content on R1.
+        let sr2 = Register::from_u16(instr & THREE_BIT_MASK)?;
+        regs[dr] = regs[sr1] & regs[sr2];
+    }
+
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Changes the PC register if the Cond register is set to the
+/// flag that is selected on the encoding of the instruction
+///
+/// ### Arguments
+///
+///
+/// - `instr`: An u16 that has the encoding of the whole instruction to execute.
+/// - `regs`: A Registers struct that handles each register.
+pub fn branch(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    // Get the PCOffset9 section
+    let mut pc_offset = instr & NINE_BIT_MASK;
+    pc_offset = sign_extend(pc_offset, 9)?;
+    // Get the Condition Flag and check if it is the same
+    // as the one selcted on the instruction
+    let cond_flag = (instr >> 9) & THREE_BIT_MASK;
+    let coincides = cond_flag & regs[Register::Cond];
+    if coincides != 0 {
```

Comment:
If the idea is to check that the condition flag and the corresponding register are the same, I think this should be like this:

```suggestion
    if cond_flag == regs[Register::Cond] {
```

Reply:
In this case, the branch instruction can have a combination of flags, and if any of those is met, you jump. So it's correct to do a bitwise and here.